pkg/controller/volume/reconciler/reconciler.go

```
@@ -75,44 +76,56 @@ func (rc *reconciler) Run(stopCh <-chan struct{}) {
 						glog.Errorf("Unexpected error actualStateOfWorld.MarkDesireToDetach(): %v", err)
 					}
-					if timeElapsed > rc.maxSafeToDetachDuration {
-						glog.V(5).Infof("Triggering DetachVolume for volume %q to node %q. Volume is not safe to detach, but max wait time expired.", attachedVolume.VolumeName, attachedVolume.NodeName)
-						rc.attacherDetacher.DetachVolume(&attachedVolume, rc.actualStateOfWorld)
+					if timeElapsed > rc.maxWaitForUnmountDuration {
+						glog.V(5).Infof("Attempting to start DetachVolume for volume %q to node %q. Volume is not safe to detach, but maxWaitForUnmountDuration expired.", attachedVolume.VolumeName, attachedVolume.NodeName)
+						err := rc.attacherDetacher.DetachVolume(attachedVolume, rc.actualStateOfWorld)
+						if err == nil {
+							glog.Infof("Started DetachVolume for volume %q to node %q. Volume was not safe to detach, but maxWaitForUnmountDuration expired.", attachedVolume.VolumeName, attachedVolume.NodeName)
+						}
 					}
 				}
 			}
 		}
+
+		// Ensure volumes that should be attached are attached.
+		for _, volumeToAttach := range rc.desiredStateOfWorld.GetVolumesToAttach() {
+			if rc.actualStateOfWorld.VolumeNodeExists(
+				volumeToAttach.VolumeName, volumeToAttach.NodeName) {
+				// Volume/Node exists, touch it to reset detachRequestedTime
+				glog.V(12).Infof("Volume %q/Node %q is attached--touching.", volumeToAttach.VolumeName, volumeToAttach.NodeName)
+				_, err := rc.actualStateOfWorld.AddVolumeNode(
+					volumeToAttach.VolumeSpec, volumeToAttach.NodeName)
+				if err != nil {
+					glog.Errorf("Unexpected error on actualStateOfWorld.AddVolumeNode(): %v", err)
+				}
+			} else {
+				// Volume/Node doesn't exist, spawn a goroutine to attach it
+				glog.V(5).Infof("Attempting to start AttachVolume for volume %q to node %q", volumeToAttach.VolumeName, volumeToAttach.NodeName)
+				err := rc.attacherDetacher.AttachVolume(volumeToAttach, rc.actualStateOfWorld)
+				if err == nil {
```

warning: swallowing error